### PR TITLE
#184581204 Users Public Chat Space

### PR DESCRIPTION
**Pull request by GasanaJr**
#### What does this PR do?
This PR adds a chat space to the clients of the E-commerce website
#### Description of Task to be completed?
```
GIVEN the user is online
WHEN they open the chat tab
THEN they should be able to read all of the past chats

GIVEN the user is online
WHEN they open the chat tab
THEN they should be able to post a public chat

GIVENthe user
WHEN they post a chat
THEN it should be able to be echoed to all online users in realtime.

```
#### How should this be manually tested?
- clone this repo `git clone https://github.com/atlp-rwanda/e-comm-team-amigos-bn`
- switch your working directory `cd e-comm-team-amigos-bn`
- create an `.env` and update it with `.env-example`
- install all dependencies by running `npm install`
- in your terminal run `npm start`
- create a sample `front-end` for `login` and `chat` example can be found @ `https://github.com/GasanaJr/Sample-Chat-Application/tree/master` in the `home.ejs` file found in the `views` directory  
- create/ login two or more users 
- Send a message from one user and it will be `broadcasted` to all others in `real-time`
#### Test the endpoints using Postman:
- GET all messages `/chat`
#### Any background context you want to provide?
- PostgreSQL installed in your local machine
- Node JS installed in your local machine
- Basic understanding of  `Sockets`
#### Screenshots
- Screenshot 2023-03-22 at 12 37 18 AM
- Screenshot 2023-03-22 at 11 32 43 AM
#### What are the relevant pivotal tracker stories?
Story ID: [#184581204](https://www.pivotaltracker.com/n/projects/2632829/stories/184581204)

### CODE REVIEW

**Comment by Emile-SHUMBUSHO** [comment]
I'm suggesting to name files inside controller without .controller since we are already in the controller directory. I could maintain consistance of our project structure.

**Comment by GasanaJr** [comment]
> I'm suggesting to name files inside controller without .controller since we are already in the controller directory. I could maintain consistance of our project structure.

I understand the concern, but wouldn't it be better if we stick with the .controller just as we do in tests and documentation

File: src/controllers/chat.controller.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+const getChat = async(req,res) => {
+    try {
+        if (req.query.df) {
+            throw new Error("No Parameters required");
+        }
+        const chats = await models.Chat.findAll()
+        return res.status(200).json({"Chats": chats});
+    } catch (error) {
+        console.log(error)
```
Comment (nshutijonathan) [inline src/controllers/chat.controller.js]: remove `console.log`

**Review by nshutijonathan — CHANGES_REQUESTED** [review]

File: src/controllers/chat.controller.js (hunk 1 above)
Comment (GasanaJr) [reply src/controllers/chat.controller.js]: Ooohhh, i was using this to read errors on the console. Sorry I forgot to change it.

**Review by wilbrord2 — APPROVED** [review]
Great work

**Review by AbdulKhaliq59 — APPROVED** [review]

**Review by khevin2 — APPROVED** [review]

**Review by Mystica52 — APPROVED** [review]

**Review by nshutijonathan — APPROVED** [review]
LGTM

File: src/controllers/chat.controller.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        return res.status(200).json({"Chats": chats});
+    } catch (error) {
+        return res.status(500).json({error: error});
+    }
+}
+
+export default {
+    getChat,
+    chat
+}
```
Comment (Habinezajanvier) [inline src/controllers/chat.controller.js]: new line needed here

File: src/database/migrations/20230317213206-create-chat.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      updatedAt: {
+        allowNull: false,
+        type: Sequelize.DATE
+      }
+    });
+  },
+  async down(queryInterface, Sequelize) {
+    await queryInterface.dropTable('Chats');
+  }
+};
```
Comment (Habinezajanvier) [inline src/database/migrations/20230317213206-create-chat.js]: new line needed here

File: src/docs/chat.doc.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+*          summary: This helps to get all chats of all clients from database.
+*          description: This will add value to the website since Clients will be able to chat on the website
+*          responses:
+*                  200:
+*                     description: Chats retrieved successfully!
+*                  401:
+*                     description: Unauthorized
+*                  500:
+*                     description: Internal server error
+*/
```
Comment (Habinezajanvier) [inline src/docs/chat.doc.js]: new line needed here

File: src/helpers/verifySocketToken.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+          }
+          return resolve(decoded);
+        }
+      });
+    });
+  }
+
+export default {
+    verifySocketToken
+}
```
Comment (Habinezajanvier) [inline src/helpers/verifySocketToken.js]: new line

**Pull request merged**